borsh/src/de/mod.rs

```diff
@@ -516,6 +511,70 @@ pub mod hashes {
             Ok(result)
         }
     }
+
+    #[cfg(feature = "de_strict_order")]
+    pub mod strict_order {
+        use crate::BorshDeserialize;
+        use crate::__maybestd::collections::{HashMap, HashSet};
+        use crate::__maybestd::io::{Error, ErrorKind, Read, Result};
+        use crate::de::ERROR_WRONG_ORDER_OF_KEYS;
+        use core::cmp::Ordering;
+        use core::hash::{BuildHasher, Hash};
+
+        impl<T, H> BorshDeserialize for HashSet<T, H>
+        where
+            T: BorshDeserialize + Eq + Hash + PartialOrd,
+            H: BuildHasher + Default,
+        {
+            #[inline]
+            fn deserialize_reader<R: Read>(reader: &mut R) -> Result<Self> {
+                let vec = <Vec<T>>::deserialize_reader(reader)?;
+                //TODO: replace with `is_sorted` api when stabilizes https://github.com/rust-lang/rust/issues/53485
+                for index in 0..(vec.len() - 1) {
+                    let (a, b) = (&vec[index], &vec[index + 1]);
```

Comment:
This will crash on len() == 0 (please add a test if there is no test catching it). I wish [`array_windows()`](https://doc.rust-lang.org/stable/std/primitive.slice.html#method.array_windows) method is stable... [window()](https://doc.rust-lang.org/stable/std/primitive.slice.html#method.windows) is useful, but it would require extra unpacking, which makes it questionable improvement:

```suggestion
                for pair in vec.windows(2) {
                    if let [a, b] = pair {
                    }
```